migration/idempotent/008-tracing-functions.sql

```diff
@@ -589,3 +589,119 @@ LANGUAGE sql;
 GRANT EXECUTE ON FUNCTION ps_trace.delete_all_traces() TO prom_admin;
 COMMENT ON FUNCTION ps_trace.delete_all_traces IS
 $$WARNING: this function deletes all spans and related tracing data in the system and restores it to a "just installed" state.$$;
+
+CREATE OR REPLACE FUNCTION ps_trace.tag_map_in(cstring)
+RETURNS ps_trace.tag_map
+LANGUAGE internal
+IMMUTABLE PARALLEL SAFE STRICT
+AS $function$jsonb_in$function$
+;
+
+CREATE OR REPLACE FUNCTION ps_trace.tag_map_out(ps_trace.tag_map)
+RETURNS cstring
+LANGUAGE internal
+IMMUTABLE PARALLEL SAFE STRICT
+AS $function$jsonb_out$function$
+;
+
+CREATE OR REPLACE FUNCTION ps_trace.tag_map_send(ps_trace.tag_map)
+RETURNS bytea
+LANGUAGE internal
+IMMUTABLE PARALLEL SAFE STRICT
+AS $function$jsonb_send$function$
+;
+
+CREATE OR REPLACE FUNCTION ps_trace.tag_map_recv(internal)
+RETURNS ps_trace.tag_map
+LANGUAGE internal
+IMMUTABLE PARALLEL SAFE STRICT
+AS $function$jsonb_recv$function$
+;
+
+DO
+$do$
+
+/* Create subscript_handler function for pg v14+.
+ * For pg v13 jsonb type doesn't have a subscript_handler function so it shall
+ * be omitted.
+ */
+DECLARE
+    _pg_version int4 := current_setting('server_version_num')::int4;
+BEGIN
+    SET search_path = pg_catalog;
+    IF (_pg_version >= 140000) THEN
+        CREATE OR REPLACE FUNCTION ps_trace.tag_map_subscript_handler(internal)
+                RETURNS internal
+                LANGUAGE internal
+                IMMUTABLE PARALLEL SAFE STRICT
+                AS $f$jsonb_subscript_handler$f$;
+
+        IF (SELECT
+                typsubscript = 'ps_trace.tag_map_subscript_handler'::regproc
```

Comment:
This will alter the type if the type already has the subscript, right? Don't we want to alter the type if the type DOES NOT have the subscript? I think we should check if the typsubscript is null or something, right?

Reply:
Yes, but my assumption was that this is essentially a no-op if it already has one, so I didn't bother with the check. Fixed now.